### PR TITLE
Global context

### PR DESCRIPTION
**Pull request by hs-lsong**
JinJava can be a singleton and shared by multiple threads.  The JinJava rendering context is scoped, but the root of the context -- the global context -- is shared. However, the context should not be shared, especially they can be modified during the rendering. To alleviate the issue,

1) we add a read only accessor, jinjava.getGlobalContextReadOnly(),
2) proxy the registering functions, so that a caller does not have to access the globalContext directly.
3) make a copy of the global context as the root of scoped context for each rendering.

Drawback is once the JinJava's `renderForResult` is started, updating the global content (registering new filter, etc) has no effect. It might actually be desired.

### CODE REVIEW

File: src/main/java/com/hubspot/jinjava/Jinjava.java [hunk 1]
```diff
@@ -123,6 +127,10 @@ public Context getGlobalContext() {
     return globalContext;
   }
 
+  public Context getGlobalContextReadOnly() {
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/Jinjava.java:130]:
```suggestion
  public Context getGlobalContextCopy() {
```

Comment (hs-lsong) [reply src/main/java/com/hubspot/jinjava/Jinjava.java:130]: I used to use `getGlobalContextCopy`, thought the readOnly one is better :).

**Comment by Joeoh** [comment]
Thanks for this. Looks like it will resolve the issue I introduced [here](https://github.com/HubSpot/jinjava/pull/438). Was that your intention or did I just get lucky here? 

I was also going to introduce a flag on Context that would identify it as the global. Not sure if you might consider that useful here so we can easily avoid adding to the context if we know its the global one.

**Comment by Joeoh** [comment]
Created PR to add global context flag to this: https://github.com/HubSpot/jinjava/pull/448

**Comment by Joeoh** [comment]
Do you plan on merging this soon?

**Comment by hs-lsong** [comment]
The branch deploy was good. Planing to merge early next week. I wanted to do a bit more debugging before we stop sharing the global context.

**Comment by Joeoh** [comment]
Sounds good! I'll wait for your changes before merging mine. They give me an extra layer of safety for my changes.

**Pull request merged**